Raw binary blob support

technikelly opened the pull request:
## Checklist

### Which kind of PR do you create?

- [x] This PR only contains minor fixes.
- [ ] This PR contains major feature update.
- [ ] This PR introduces a new function/api for Qiling Framework.

### Coding convention?

- [x] The new code conforms to Qiling Framework naming convention.
- [x] The imports are arranged properly.
- [x] Essential comments are added.
- [x] The reference of the new code is pointed out.

### Extra tests?

- [ ] No extra tests are needed for this PR.
- [x] I have added enough tests for this PR.
- [ ] Tests will be added after some discussion and review.

### Changelog?

- [ ] This PR doesn't need to update Changelog.
- [x] Changelog will be updated after some proper review.
- [ ] Changelog has been updated in my PR.

### Target branch?

- [x] The target branch is dev branch.

### One last thing

- [x] I have read the [contribution guide](https://docs.qiling.io/en/latest/contribution/)

-----

## Overview

This PR introduces a small code change that allows users to specify BLOB_RAW in the profile, instructing the Qiling loader not to map any memory segments. This functionality is particularly useful for emulating small code snippets or bare-metal code without an operating system, offering greater customization and flexibility. Users can now define all memory mappings directly within their Qiling scripts, tailoring the environment precisely to their use case. 

With this change, it becomes easier to port existing Unicorn scripts to Qiling and take advantage of additional features Qiling provides, such as debugging support and code coverage generation. Users can simply replace Unicorn API calls with equivalent Qiling API calls to benefit from these enhancements.

## Notes

The added unit tests are currently disabled because the test binary has not been merged to rootfs. The added example code will also not work until the test binary has been merged into rootfs. Corresponding rootfs PR - https://github.com/qilingframework/rootfs/pull/36

technikelly committed:
adding support for raw binary blobs

elicn commented:
Hi @technikelly, welcome and thank you for your contribution.

Though the motivation behind this PR does resonate with me, I believe it introduces a special use case that could be covered by the existing Blob loader with minimal changes. While providing a convenient method for their users, special use cases tend to serve a very focused or limited purpose and introduce more complexity to the framework as it needs to support these cases on various components.

This is what I've picked up in my review:
 - The new approach uses `binary_name` key in profile, which is inconsistent with Qiling usage model. Blob is assumed to be provided directly as `bytes`, that is actually more "Unicorn-like"
 - Profile has a `binary_size` key which appears to be redundant, as file size is inferred when it is being read or simply by the length of code `bytes`
 - The existing Blob loader could be modified as follows to provide an equivalent functionality:
   - Treat the `heap_size` key as optional and configure the heap only if it is available. This way only the bare minimum memory is allocated, avoiding other Qiling "noise"
   - Everything else remains intact, while the user script can monkeypatch details as needed (e.g.: patch module name in loader, unmap unnecessary memory areas, etc.)

Also, please note that we cannot accept proprietary code with copyrights.
Please consider this feedback and let me know if you have any concerns.

[BTW - you may want to consider [udbserver](https://github.com/bet4it/udbserver) as a simpler method to debug Unicorn scripts]

technikelly commented:
Hi @elicn, thank you for getting back to me so quickly. I appreciate your thorough explanation of your reasoning and I fully agree that this could be achieved with a much smaller code change to qiling and monkeypatching the user script. 

I'll go ahead and make the change you suggested - of `heap_size` being configurable within the blob loader.

Regarding the copyrights, the code is not proprietary. Cisco requires a copyright notice on open source contributions, but the code ultimately falls under the license of the overall project which in this case is GPLv2. With that additional information is it still an issue to have copyright notices included with contributions to qiling? 

Thanks for the tip about udbserver - it definitely is simpler, but having code coverage in addition to debug is really nice.

technikelly committed:
minimizing blob changes

technikelly committed:
updating copyright statements with license info

technikelly committed:
update description

xwings commented:
> Hi @elicn, thank you for getting back to me so quickly. I appreciate your thorough explanation of your reasoning and I fully agree that this could be achieved with a much smaller code change to qiling and monkeypatching the user script. 
> 
> 
> 
> I'll go ahead and make the change you suggested - of `heap_size` being configurable within the blob loader.
> 
> 
> 
> Regarding the copyrights, the code is not proprietary. Cisco requires a copyright notice on open source contributions, but the code ultimately falls under the license of the overall project which in this case is GPLv2. With that additional information is it still an issue to have copyright notices included with contributions to qiling? 
> 
> 
> 
> Thanks for the tip about udbserver - it definitely is simpler, but having code coverage in addition to debug is really nice.

Hi,

Thank you for PR and also welcome to Qiling Framework.

We do have the "tradition" to include author (PR) during the early days. Along the way we found out the list will get longer and longer in the specific file.

Later, we decided update changelog and put the names in sparate for easier maintainance. I hope you understand where we come from.

elicn commented on examples/hello_arm_blob_raw.py:
```diff
@@ -0,0 +1,92 @@
+##############################################################################
```
Examples are an essential resource for beginners to learn how a certain function of Qiling works. For an example to be effective, it has to describe clearly what is does and how it does it, otherwise the example does not add any value to the reader.

This example file seems to be the very specific case you have been working on, so you'll need to ask yourself what is in there for other Qiling users. If you think it is valuable, kindly revise the example file and add comments and documentation.

technikelly replied:
Good point, thank you for that feedback. This example code is a generic fabricated example with the goal of being simple and  having multiple code paths that could be explored and therefore highlighted with code coverage. I believe a beginner would not know that they could emulate a small code snippet in this way. Although, if you feel that the example isn't beneficial, it is ultimately your decision. Let me know if you would prefer me to remove it altogether.

elicn commented on examples/hello_arm_blob_raw.py:
```diff
@@ -0,0 +1,92 @@
+    expected_checksum_python &= 0xFF # Ensure it's a single byte
+    return expected_checksum_python
+
+def unmapped_handler(ql, type, addr, size, value):
```
Please use annotations where applicable.
Annotations have to be compliant to Python 3.8

technikelly replied:
apologies, addressed in https://github.com/qilingframework/qiling/pull/1583/commits/5ec4a3b57a0cdf6e638026e04ab94cc60b1aabb5

elicn commented on examples/hello_arm_blob_raw.py:
```diff
@@ -0,0 +1,92 @@
+
+def unmapped_handler(ql, type, addr, size, value):
+
+    print(f"Unmapped Memory R/W, trying to access {hex(size)} bytes at {hex(addr)} from {hex(ql.arch.regs.pc)}")
```
While still possible, `hex` is considered obsolete. Could replace this with:
```python
print(f"Unmapped Memory R/W, trying to access {size:d} bytes at {addr:#010x} from {ql.arch.regs.pc:#010x}")
```
FYI, ARM has an additional `effective_pc` method that may be used to retrieve the 'real' PC value, taking thumb mode (odd address) into consideration - just in case you need it.

technikelly replied:
good to know, thanks!

elicn commented on examples/hello_arm_blob_raw.py:
```diff
@@ -0,0 +1,92 @@
+    # Remove all images with name 'blob_code' that were created by the blob loader
+    ql.loader.images = [img for img in ql.loader.images if img.path != 'blob_code']
+    # Add image back with correct info
+    ql.loader.images.append(Image(ql.loader.load_address, ql.loader.load_address + ql.os.code_ram_size, 'example_raw.bin'))
```
Better avoid hardcoding filenames

technikelly replied:
addressed in - https://github.com/qilingframework/qiling/pull/1583/commits/5ec4a3b57a0cdf6e638026e04ab94cc60b1aabb5

elicn commented on qiling/loader/blob.py:
```diff
@@ -31,8 +36,12 @@ def run(self):
-        heap_size = int(self.ql.os.profile.get("CODE", "heap_size"), 16)
-        self.ql.os.heap = QlMemoryHeap(self.ql, heap_base, heap_base + heap_size)
+        # if heap_size is defined, create the heap
+        try:
```
Better use this opportunity to move this to Blob OS, as the `FIXME` comment above suggests.
I would also suggest reading `heap_base` along with `heap_size` to keep consistency with other use cases.

Also note that we have a better method to extract integer values from config:
```python
self.ql.os.profile.getint("CODE", "heap_size")
```
It already takes care of whatever representation it is there (not assuming hex specifically).

Putting all of this together, we can remove the two `FIXME` comments above.

technikelly replied:
oh good to know, I do see that used in other loaders now that you mention it. But good point, I did address those FIXME's in https://github.com/qilingframework/qiling/pull/1583/commits/5ec4a3b57a0cdf6e638026e04ab94cc60b1aabb5 as a result I modified uboot_bin.ql to contain a value for `heap_address`

elicn commented on tests/test_blob.py:
```diff
@@ -82,6 +86,101 @@ def partial_run_init(ql: Qiling):
+            return expected_checksum & 0xFF
+
+        # Test cases with descriptions
+        test_cases = {
```
Though I understand where this is coming from [de-duplicating test code], it is a `unittest` bad practice to bundle multiple test cases in one method: if one of these tests fails, everything else is also considered as a failure.

Please create dedicated `test_*` methods for each test case (not sure that we need all, though, after all we test Qiling functionality, not the binary) and extract the ql prep code to a method all test cases can call.

technikelly replied:
Sure thing, thanks for pointing that out

elicn reviewed (changes requested):
@technikelly, thanks for the fixes.
Please see my review comments inline.

I believe we would need to get rid of those repeating copyright notices, as @xwings suggested we can include your name and your employer's name in the contributors list.

technikelly committed:
addressing review comments, removing copyright notices

technikelly commented:
Thank you @elicn and @xwings for explaining about the issue with copyright notices. I was able to get an exception, so I've removed them. I've also implemented the changes requested by @elicn (thank you for the review!). As it stands now, the changes in this PR consist of:

* moving the heap creation to the blob OS module, and making that heap creation optional based on the existence profile values `heap_address` and `heap_size`
* Test code 
* Example showcasing the modifications needed to enable code coverage

As stated in the comments, let me know if you would prefer not to include an example with this PR. 

I also wanted to point out that I modified the two instances of uboot_bin.ql so that they contain `heap_address` so that the code change wouldn't break the uboot example or test

elicn commented:
Everything looks good to me. Thanks @technikelly.
We'll merge the changes to rootfs first to allow the tests here to pass.

technikelly committed:
enabling test and updating rootfs submodule

technikelly commented:
Great, thank you @elicn! I just enabled the test and updated rootfs.

elicn commented:
@xwings, pending your review.

xwings reviewed (approved):
Looks good to me.

xwings commented:
Merged and thanks! Again, welcome to Qiling Framework
